service-proxy.js

```diff
@@ -950,7 +949,44 @@ function reapSinglePeer(hostPort) {
         self.deletePeerIndex(serviceName, hostPort);
     }
 
-    self.channel.peers.delete(hostPort);
+    // TODO: info log/stat
+
+    peer.drain({
+        reason: 'reaped for expired advertisement',
+        direction: 'both',
```

Comment:
Currently there is a "bug" in connection.js

 - on init response flushed
 - get peer for `hostPort`
 - Ignore the fact that the peer might be draining or closing
 - add the incoming connection to the peer
 - #leak 
 - profit

{ direction: 'both', onlyAccept: true } semantics would be something like "drain all connections right now; but if there are any incoming connections during the drain timeout add them to the peer"

We would also need to do a "peer.hasConnections()" in the drain callback check to only shutdown the peer if there were no new incoming connections in the 30 second window

#compexity #:(